Proposal: display Style Specification as multiple pages

katydecorah opened the pull request:
This PR adds on to the work happening in #91 .

@mapbox/docs per chat, we discussed displaying the [Style Specification](https://docs.mapbox.com/mapbox-gl-js/style-spec/) as multiple pages.

## Why?

* The current Style Specification page is long, can be difficult to digest, and difficult to scan to find what you're looking for. 
* The sidebar isn't optimized to handle the amount of items on the page.
* Onsite docs search isn't as effective on long pages as you still need to scan or ctrl+f to find what you were initially searching.

## Proposal

This proposed PR will display each section as its own page and use the `NavigationAccordion` component that we use on other docs sites. Each page will have a multi-level table of contents within the sidebar making it easier to jump to headings and properties within each page.

Pros to multiple pages:
* Onsite search will work better.
* Easier to navigate and view from sidebar.
* Reduce complexity of h2/h3 section components. We added styling and component to better define each section on such a long page, but we're able to remove them since each page will be more focused.
* Faster page load times.

Cons to multiple pages:
* **Loose ability to ctrl+f search.** However, now that we have onsite search this is less of an issue. Onsite search for this page will also work better as content is broken into pages rather than the user landing on the original long page. (Note: we had a similar concern about losing the ability to ctrl+f on /api, but since we broken the site into pages last year, we have not received any critical feedback regarding the change.)
* ~~**External links to this site using hashes (to headings) will not work unless we update them all.** The links won't be broken, but the user will land on the main /style-spec page and need to reorient themselves to the desired section. The docs team can update all docs.mapbox.com sites, but this does not account for other links out of our control.~~ This is no longer an issue with the introduction of a JS-redirect in https://github.com/mapbox/mapbox-gl-js-docs/pull/189/commits/b8e4c4cb7da2838a1d4ab829fd191059fab783fe

## Comparisons


Status | Page
---|---
Current | ![image](https://user-images.githubusercontent.com/2180540/71900295-87356680-312b-11ea-8467-15d9aeb7d9d3.png) 
Proposed | ![image](https://user-images.githubusercontent.com/2180540/71900257-708f0f80-312b-11ea-81df-ba9b52822a54.png)

Status | Page
---|---
Current | ![image](https://user-images.githubusercontent.com/2180540/71900348-a7652580-312b-11ea-9f61-67d75ca6f044.png)
Proposed | ![image](https://user-images.githubusercontent.com/2180540/71900317-987e7300-312b-11ea-8e16-fb3e29039be2.png)



## Changes in this PR

* Updated style-spec-navigation.js. This file powers the current sidebar and will continue to do so.
    * It’s not currently possible to build the sidebar programmatically given that the pages are built from different sources and Batfish ingests these data sources at different times. I was able to pull in data via source where possible to better keep the sub navigation items in sync. 
    * We’ll need to keep this up-to-date if we change or add sections. This isn't a changed behavior, but we should make sure this is documented. It may also be helpful to add a comment to the top of each page to serve as a reminder that if headings are changed on the page, then they should be updated in `style-spec-navigation.js`.
    * I turned off the bolding of active sidebar titles on the sidebar scroll-spy, which is a feature in NavigationAccordion. Due to the disparate data sources, the bolding wasn't consistent across pages.
* Removed components we no longer need:
  * Removed SectionH2, SectionH3, SectionH4. We now use h2 or h3 for headings.
  * Removed InnerSection, markdown-mini-shell.js and section-intro.js, no longer needed.
* Added components to centralize codebases:
  * Created `Items` component to generate each `entries` item.
  * Created `Property` component that holds styles for the clickable, monospaced headings.
  * Created `Subtitle` component that holds the italicized gray metadata for each property.
  * Used `React.Fragment` in place of `<div>` where possible to reduce generation of extra divs. (Smaller chunks!)
* Created `ExpressionReference` component since this was a rather large JSX code block, it may be easier to parse as a separate component rather than embedded in the markdown.
* Combined other-filter.md into other.md page to follow the patterns of other sections where one page = one section.
* Created two new headings in /layers (the properties were otherwise lacking hierarchy to make the sidebar effective):
    * Layer properties
    * Layout and Paint Properties
* Updated all hash links to their new page paths.

colleenmcginnis commented:
I am quite excited about this proposal @katydecorah! A few initial thoughts before digging into the diff:  

>Loose ability to ctrl+f search. 

@danswick and I have both experienced buggy behavior when trying to use ctrl+f on long pages like the Style Spec and the API reference in this repo so we might be losing something that doesn't work consistently anyway. 🙂 

>Cons to multiple pages:

Another con is the redirect situation. Also similar to the API documentation rebuild, we have many many links out there pointing to a specific hash on the single Style Spec page. Since it's not possible to redirect from a specific hash to a new page, we would need to update all links on pages we own and there will likely be many links out in the wild that will just drop users on the intro page. I think a case can be made that the longterm improvements to user experience outweigh the potential confusion caused by this issue (imo that was the case with API documentation), but it should be factored into our decision to move forward here! 

(We _may_ also be able to get away with using more `##` and `###` instead of `h2` and `h3` with custom `id`s because all the existing links out in the wild that include hashes will be incorrect anyway.)

>Created two new headings in /layers (the properties were otherwise lacking hierarchy to make the sidebar effective):
>- Layer properties
>- Layout and Paint Properties

My gut tells me the layer types should be the `h2`s (for findability and to align with other sections like "Sources"), but that would require some reframing in this section. Maybe @danswick or the GL JS team will have a strong opinion / suggestion.

![image](https://user-images.githubusercontent.com/10479155/71941606-3d6e6f80-316f-11ea-9079-796c4196b109.png)

katydecorah commented:
> Another con is the redirect situation. 

@colleenmcginnis you are so right, I forgot about this one. I also agree that  the longterm improvements to user experience outweigh the potential confusion caused by this issue. Especially since no links will be broken, the user will need to reorient themselves via the sidebar or search once they land on the introduction Style Spec page. I'll add this to the original comment to keep this all together.

> (We may also be able to get away with using more ## and ### instead of h2 and h3 with custom ids because all the existing links out in the wild that include hashes will be incorrect anyway.)

Yes, I would prefer using markdown headings. Good point! If we feel ok about the previous point, we should rework the hashes to remove their prefixes to allow for using markdown headings.

> My gut tells me the layer types should be the h2s (for findability and to align with other sections like "Sources"), but that would require some reframing in this section

This is the feedback I was hoping for - I'll brainstorm on how to better organize this section.

katydecorah commented:
Follow-up on:

> Another con is the redirect situation.

@colleenmcginnis https://github.com/mapbox/mapbox-gl-js-docs/pull/189/commits/b8e4c4cb7da2838a1d4ab829fd191059fab783fe contains a function that will redirect any eligible hashes that land on the main /style-spec page to their new page.

* The redirect function (it's called on [page_shell.js](https://github.com/mapbox/mapbox-gl-js-docs/blob/b8e4c4cb7da2838a1d4ab829fd191059fab783fe/docs/components/page_shell.js#L33-L40)): https://github.com/mapbox/mapbox-gl-js-docs/blob/b8e4c4cb7da2838a1d4ab829fd191059fab783fe/docs/util/style-spec-redirect.js
* A test to assert the redirect function: https://github.com/mapbox/mapbox-gl-js-docs/blob/b8e4c4cb7da2838a1d4ab829fd191059fab783fe/test/style-spec-redirect.test.js

I think this is a promising solution!

katydecorah commented:
@asheemmamoowala @mourner @chloekraw This PR scopes out what turning the Style Specification page into many pages could look like. **Do you have any concerns with turning the Style Specification into multiple pages?**

We've outlined the pros/cons of turning Style Spec into multiple pages in the [original post of this PR](https://github.com/mapbox/mapbox-gl-js-docs/pull/189#issue-360075548). 

This branch is also on staging if you'd like to interact with the proposed Style Spec pages.

asheemmamoowala commented:
Overall I think the breakdown of pages feels right.

1. Style Spec home page: This section feels like it could use additional detail on the structure of a stylesheet and links to the high level sections.

2. The styling of source/layer types feels too light. It doesn't draw attention as an important element, the way that Expression page headings do.

  <details>
  <summary>Screenshot</summary>

  ![image](https://user-images.githubusercontent.com/11619675/72322752-f8a97380-365b-11ea-929a-79a21e1c3f2d.png)

</details>

The whitespace/negative space on these pages also seems a bit odd. It feels like sections need better visual separation and reduced vertical padding (?), but there's probably better ways to address this as well.

3. The **Expressions** page is still quite long. Would it make sense to separate the `Expressions reference` into its own page? We aren't doing that for individual source and layer types so it may feel consistent, but it is a page that is already quite long. 

4. The **Other** page at the bottom could also be broken into individual pages for `Function`s and `Deprecated Filter Syntax`. In fact both of these sections could be marked as deprecated.

5. On the **Glyphs** page there seems to be a small bug with inline code formatting:
  <details> <summary>screenshot</summary>

![image](https://user-images.githubusercontent.com/11619675/72323421-90f42800-365d-11ea-8ae7-0053d36c70a7.png)

 </details>

danswick commented:
@katydecorah I added some copy in 9c02dd7ee7d822664b3b5d92f14346cdcf40252e, but I'm really not sure if it's useful. As we talked about in chat, the new site organization and the accordion nav already provide a clearer overview of the style document's structure, so I was hesitant to describe that structure again in narrative form here. Very open to suggestions and ideas if you have any!

katydecorah commented:
@danswick I think the intro is great! What do you think about adding some links for "root properties", "layers", "sources", etc?

katydecorah commented:
@asheemmamoowala thank you for your feedback! We're still working through some of the items, but here a few follow ups:

> The styling of source/layer types feels too light. It doesn't draw attention as an important element, the way that Expression page headings do.

We pushed some styling updates to make the styling a little more consistent and tighter among sections.

> The Expressions page is still quite long. Would it make sense to separate the Expressions reference into its own page? We aren't doing that for individual source and layer types so it may feel consistent, but it is a page that is already quite long.

We investigated splitting these pages, but since Expression reference content made up for a majority of the content, not much was gained from splitting the pages. Instead, we can rework the sidebar to use the same sidebar content that's on production - listing just the Expression reference elements. This makes accessing the content much more approachable.

> The Other page at the bottom could also be broken into individual pages for Functions and Deprecated Filter Syntax. In fact both of these sections could be marked as deprecated.

Since this is deprecated content, we'd like to keep the "Other" page and add a legacy tag and notes to each section indicating its status. After this PR merges, we should discuss internally the future of handling deprecated content for GL JS and then we can follow-up with the "Other" page as needed to avoid blocking this PR. (With /api we move legacy content off the main sidebar and into a legacy directory that's only reachable via link -- not sure if this pattern makes sense for this content and site, but it would be good to have a process that we can follow when features are deprecated.)

> On the Glyphs page there seems to be a small bug with inline code formatting:

✅ fixed!

katydecorah commented on docs/pages/style-spec/other.md:
```diff
@@ -4,68 +4,55 @@ id: other
-A set of one input value and one output value is a "stop." Stop output values must be literal values (i.e. not functions or expressions), and appropriate for the property. For example, stop output values for a `fill-color` function property must be [colors](#types-color).
-
-{{</SectionH4>}}
+## Function
```
To do: add a legacy note here.

katydecorah replied:
@danswick do you have any suggestion on how to frame this legacy note?

danswick replied:
Maybe something like this?

> Starting in [Mapbox GL JS v0.41.0](https://github.com/mapbox/mapbox-gl-js/blob/master/CHANGELOG.md#0410-october-11-2017), [property expressions](/mapbox-gl-js/style-spec/expressions) became the preferred method for styling features based on zoom level or the feature's properties. Zoom and property functions are still supported, but will be phased out in a future release.

previous: https://github.com/mapbox/help/issues/2257

katydecorah replied:
That's perfect, thank you!

katydecorah commented on docs/pages/style-spec/layers.md:
```diff
@@ -56,38 +47,87 @@ Content in this section is generated directly using the Mapbox Style
 -->
+
+
+## background
```
Note: for scroll-spy to work properly for these sections we need the markdown H2s

asheemmamoowala commented:
Thanks @katydecorah looks great!

> We pushed some styling updates to make the styling a little more consistent and tighter among sections.

The amount of white space between individual properties on the source and layer pages still feels large. Is this by design? Can it be reduced or maybe style those headers to reduce the feeling of empty space?

chloekraw commented:
👍 overall I'm good with splitting the style spec docs into multiple pages. thanks for the work! I think this will be an improvement.

danswick commented:
re: the whitespace between properties, I personally think the additional space is an improvement. This content is very dense and I've always found that I get lost easily when scrolling.

katydecorah commented:
> The amount of white space between individual properties on the source and layer pages still feels large. Is this by design? Can it be reduced or maybe style those headers to reduce the feeling of empty space?

@asheemmamoowala I pushed a commit to making the spacing between individual properties closer together.

asheemmamoowala commented:
@katydecorah 🙇 for the spacing changes and the `legacy` badge. Looks 💯 !!

danswick commented on docs/components/style-spec/property.js:
```diff
@@ -0,0 +1,24 @@
+export default class Property extends React.Component {
+    render() {
+        return (
+            <a
```
What do you think about this component accepting a prop for heading level? I feel like we could be more semantic and better for screen reader accessibility to make these into headings.

katydecorah replied:
I agree, I'd prefer to use headings. The problem is that the formatting would look different between pages since some use this component in the place of an H2 and others an H3. I wonder if we should still use headings, but just override the styling so they are consistent. I'll give that a try.

danswick commented on docs/util/style-spec-redirect.js:
```diff
@@ -0,0 +1,56 @@
+    }`;
+}
+
+module.exports = redirect;
```
Love this! 😍

katydecorah commented on docs/pages/style-spec/index.md:
```diff
@@ -1,6 +1,6 @@
 ---
 title: Style Specification
-id: intro
+id: order
```
copy pasta perhaps! I'll revert

katydecorah commented:
@danswick Your suggestion to pass the heading to Property was awesome! I incorporated it in https://github.com/mapbox/mapbox-gl-js-docs/pull/189/commits/00f9cb12b1320339a8c605a472df92649c400288. I think I addressed all other feedback.
